src/BackgroundPublish.h

```diff
@@ -117,78 +137,174 @@ class BackgroundPublish {
-                flags, 
-                level, 
-                std::bind(cb, instance, std::placeholders::_1, 
-                    std::placeholders::_2, std::placeholders::_3, std::placeholders::_4), 
-                context);
-}
\ No newline at end of file
+template<std::size_t NumQueues>
+void BackgroundPublish<NumQueues>::start()
+{
+    if (running) {
+        logger.warn("start() called on running publisher");
+        return;
+    }
+    running = true;
+    _thread = Thread("background_publish",
+                     std::bind(&BackgroundPublish::thread, this),
+                     OS_THREAD_PRIORITY_DEFAULT);
+}
+
+template<std::size_t NumQueues>
+void BackgroundPublish<NumQueues>::stop()
+{
+    if (!running) {
+        logger.warn("stop() called on non-running publisher");
+        return;
+    }
+    running = false;
+    _thread.join();
+    cleanup();
+}
+
+template<std::size_t NumQueues>
+particle::Error BackgroundPublish<NumQueues>::process_publish(const publish_event_t& event)
+{
+    auto promise {Particle.publish(event.event_name,
+                                   event.event_data,
+                                   event.event_flags)};
+
+    // Can't use promise.wait() outside of the application thread
+    while(!promise.isDone()) {
+        delay(1); // yield to other threads
+    }
+    auto error {promise.error()};
+
+    if(event.completed_cb != nullptr) {
+        event.completed_cb(error,
+                           event.event_name,
+                           event.event_data,
+                           event.event_context);
+    } else {
+        if (error != particle::Error::NONE) {
+            // log error if no callback is used
+            logger.error("publish failed: %s", error.message());
+        }
+    }
+
+    return error;
+}
+
+template<std::size_t NumQueues>
+void BackgroundPublish<NumQueues>::thread() {
+    constexpr std::size_t burst_rate {2u}; // allowable burst rate (Hz), Device OS allows up to 4/s
+    constexpr system_tick_t process_interval {1000u};
+
+    system_tick_t publish_t[burst_rate] {}; // publish time of the last (burst_rate) sends in a circular buffer
+    std::size_t i {}; // publish time of the previous (burst_rate)th send
+
+    while(running) {
+        auto now {millis()};
+        if(now - publish_t[i] >= process_interval) {
+            for(auto &queue : _queues) {
+                _mutex.lock();
+                if(!queue.empty()) {
+                    publish_t[i] = now;
+                    i = (i + 1) % burst_rate;
+                    // Copy the event and pop so the publish and wait is done without holding the mutex
+                    publish_event_t event {queue.front()};
+                    queue.pop();
+                    _mutex.unlock();
+                    process_publish(event);
+                    break;
+                }
+                _mutex.unlock();
+            }
+        }
+
+        delay(1); // force yield to processor
```

Comment:
@jlovinger-particle are we supposed to use `HAL_Delay_Milliseconds(0)` now?

Reply:
This [PR](https://github.com/particle-iot/device-os/pull/2519) fixes an issue with `delay(1)` but it is from some version of 5.x onwards.
Please change this to `delay(2)` per Andrey's advice.